b2sdk/utils/session_protocol.py

```diff
@@ -0,0 +1,102 @@
+######################################################################
+#
+# File: b2sdk/utils/session_protocol.py
+#
+# Copyright 2020 Backblaze Inc. All Rights Reserved.
+#
+# License https://www.backblaze.com/using_b2_code.html
+#
+######################################################################
+import functools
+import importlib
+import os
+from typing import (
+    Dict,
+    List,
+    NamedTuple,
+    Type,
+)
+
+
+class SessionProtocolInfo(NamedTuple):
+    module_name: str
+    class_name: str
+    env_variable: str
+
+
+# This list is to be ordered from the best to the worst.
+# If environmental variable is set, only these libraries we try to load.
+# That is, if multiple variables are set, the one from the top of the list
+# will be actually used.
+# Also, if it's impossible to import a library that was set as required,
+# it's an error.
+FROM_IMPORT_ENV_LIST = [
+    SessionProtocolInfo('b2sdk.utils.curl', 'CurlSession', 'B2_USE_LIBCURL'),
+    SessionProtocolInfo('requests', 'Session', 'B2_USE_REQUESTS'),
+]
+
+
+# TODO: use typing.Protocol when we drop 3.7
+class SessionProtocol:
+    def request(self, *args, **kwargs):
+        pass
+
+    def post(self, *args, **kwargs):
+        pass
+
+    def get(self, *args, **kwargs):
+        pass
+
+    def head(self, *args, **kwargs):
+        pass
+
+    def mount(self, *args, **kwargs):
+        pass
+
+
+class Sessions(NamedTuple):
+    enabled: List[Type[SessionProtocol]]
+    disabled: Dict[str, str]
+
+
+@functools.lru_cache(maxsize=None)
+def get_session_protocols(enable_env_checking: bool = True) -> Sessions:
+    """
+    Tries to import all available session protocols. Returns a list ordered from "the best"
+    to "the worst" and a dictionary with names of protocols that failed with a reason.
+    """
+    enabled_sessions = []
+    errors = {}
+
+    # List of protocols that we should consider.
+    enabled_env_variables = set()
+
+    if enable_env_checking:
+        for protocol_info in FROM_IMPORT_ENV_LIST:
+            env_value = os.getenv(protocol_info.env_variable)
+            # We assume that any value assigned to these variables is enough. Only empty string and None are rejected.
```

Comment:
orly? `=0` and `=false` will cause the protocol to be used?

Reply:
Fair point. Only (case-insensitive) values of `y`, `yes`, `true`, `t` and `1` will be allowed. Providing anything else or providing nothing will result in a disabled value.